src/host.rs

```
@@ -133,30 +132,90 @@ impl<T> Host<T> {
+                (*self.inner).peers,
+                (*self.inner).peerCount.try_into().unwrap(),
+            )
+        };
+
+        peers.iter_mut().map(|peer| Peer::new_mut(&mut *peer))
+    }
+
+    /// Returns an iterator over all peers connected to this `Host`.
+    pub fn peers(&self) -> impl Iterator<Item = &'_ Peer<T>> {
+        let peers = unsafe {
+            std::slice::from_raw_parts(
+                (*self.inner).peers,
+                (*self.inner).peerCount.try_into().unwrap(),
+            )
+        };
 
-        let raw_peers = unsafe { std::slice::from_raw_parts_mut((*self.inner).peers, peer_count) };
+        peers.iter().map(|peer| Peer::new(&*peer))
+    }
 
-        raw_peers.iter_mut().map(|rp| Peer::new(rp))
+    fn process_event(&'_ mut self, sys_event: ENetEvent) -> Option<Event<'_, T>> {
+        Event::from_sys_event(sys_event, self)
     }
 
     /// Maintains this host and delivers an event if available.
     ///
-    /// This should be called regularly for ENet to work properly with good
-    /// performance.
-    pub fn service(&'_ mut self, timeout_ms: u32) -> Result<Option<Event<'_, T>>, Error> {
-        // ENetEvent is Copy (aka has no Drop impl), so we don't have to make sure we
-        // `mem::forget` it later on
+    /// This should be called regularly for ENet to work properly with good performance.
+    ///
+    /// The function won't block if `timeout` is less than 1ms.
+    pub fn service(&'_ mut self, timeout: Duration) -> Result<Option<Event<'_, T>>, Error> {
+        // ENetEvent is Copy (aka has no Drop impl), so we don't have to make sure we `mem::forget` it later on
         let mut sys_event = MaybeUninit::uninit();
 
-        let res = unsafe { enet_host_service(self.inner, sys_event.as_mut_ptr(), timeout_ms) };
+        let res = unsafe {
+            enet_host_service(
+                self.inner,
+                sys_event.as_mut_ptr(),
+                timeout.as_millis() as u32,
```

I guess this is the reason the API used an `u32` before: The (possible) conversion from `Duration` to `u32` can fail (if the number of millis is too big), and this version silently drops the error instead of leaving it up to the user.

I think we should go back to using an `u32` as a parameter again, the examples can then do sth. like `Duration::from_secs(1).as_millis() as u32`, making clear that there is a cast happening.

---

As service returns a result, we could also use try_into and return an Error on failure.
Unfortunately the Error may only return a negative integer as error value, so it's not necessarily clear what happened.
Although we can of course document the integer value of that specific error code.

I generally prefer the Duration API, but can see why you are concerned about the conversion.
Let me know which option you prefer.